# 1-2-Genome-annotation-analysis
Jakke Neiro$^1$
1. Aboobaker laboratory, Department of Zoology, University of Oxford

## Contents of notebook
* 1. Introduction
* 2. Comparison of annotations
    * 2.1 Rink High-confidence vs. Neiro
* 3. GO terms for Rink high-confidence transcripts
    * 3.1 High-confidence transcripts from annotation file
    * 3.2 Transdecoder
    * 3.3 Interproscan
    * 3.4 GO terms
* 4. Coding and non-coding transcripts from new annotation
    * 4.1 Putative ncRNAs from annotation file
    * 4.2 Transdecoder
        * 4.2.1 Open reading frames
        * 4.2.2 Preparation for Interproscan
    * 4.3 Inteproscan
        * 4.3.1 Sequences longer than 100 amino acids
        * 4.3.2 Sequences shorter than 1000 amino acids
    * 4.4 Characterisation of transcripts
        * 4.4.1 Coding sequences longer than 100 amino acids
        * 4.4.2 Coding sequences shorter than 100 amino acids
        * 4.4.3 Non-coding transcripts

## Files:
* Input: stringtie_merged.gtf
* Output: interpro_rink.GO.tab, ncrna_coding_May2021.GO.tab, ncrna_new_coding_May2021.GO.tab, noncoding_final_May2021.csv

# 1. Introduction

The genome annotation generated in 1-1-Genomeannotation was used to identify new coding and non-coding transcripts. In addition, GO terms were assigned to transcripts from the Rink high-confidence transcripts. 

# 2. Comparison of annotations

First, the new annotation was compared to previous annotations. 

## 2.1 Rink High-confidence vs. Neiro

The new annotation ("Neiro") was compared to the Rink's high-confidence annotation from Planmine ("Rink High-confidence")(http://planmine.mpi-cbg.de/planmine/aspect.do?name=Gene%20Predictions) with gffcompare. 

In [ ]:
#%%bash
#cd /hydra/neiro_bam
#/hydra/software/gffcompare-0.11.5/gffcompare -r ~/sexual_genome_annotation_files/sexual_genome_annotation/smes_v2_hconf_SMESG.gtf -G stringtie_merged.gtf

The gffcompare results were investigated:

In [2]:
%%bash
cd /hydra/neiro_bam
less gffcmp.stats

# gffcompare v0.11.5 | Command line was:
#/home/likewise-open/ZOO/quee4048/software/gffcompare-0.11.5/gffcompare -r /home/likewise-open/ZOO/quee4048/sexual_genome_annotation_files/sexual_genome_annotation/smes_v2_hconf_SMESG.gtf -G stringtie_merged.gtf
#

#= Summary for dataset: stringtie_merged.gtf 
#     Query mRNAs :   91068 in   28097 loci  (82514 multi-exon transcripts)
#            (14673 multi-transcript loci, ~3.2 transcripts per locus)
# Reference mRNAs :   28173 in   22129 loci  (22798 multi-exon)
# Super-loci w/ reference transcripts:    20781
#-----------------| Sensitivity | Precision  |
        Base level:   100.0     |    81.0    |
        Exon level:    98.0     |    57.3    |
      Intron level:   100.0     |    69.8    |
Intron chain level:   100.0     |    27.6    |
  Transcript level:   100.0     |    30.9    |
       Locus level:   100.0     |    73.9    |

     Matching intron chains:   22798
       Matching transcripts:   28160
              Matching loci:   2212

In [3]:
%%bash
cd /hydra/neiro_bam
head -2 gffcmp.tracking

TCONS_00000001	XLOC_000001	SMESG000026673.1|SMEST026673001.1	=	q1:MSTRG.1|SMEST026673001.1|2|0.000000|0.000000|0.000000|1072
TCONS_00000002	XLOC_000001	SMESG000026673.1|SMEST026673001.1	m	q1:MSTRG.1|MSTRG.1.2|1|0.000000|0.000000|0.000000|1042


In [5]:
%%bash
cd /hydra/neiro_bam
head -2 gffcmp.loci

XLOC_000001	dd_Smes_g4_1[+]20346-21503	SMESG000026673.1|SMEST026673001.1	SMEST026673001.1,MSTRG.1.2,MSTRG.1.3
XLOC_000002	dd_Smes_g4_1[+]30235-31845	-	MSTRG.2.1,MSTRG.2.2


In [6]:
%%bash
cd /hydra/neiro_bam
head -2 gffcmp.stringtie_merged.gtf.refmap

ref_gene_id	ref_id	class_code	qry_id_list
SMESG000026673.1	SMEST026673001.1	=	MSTRG.1|SMEST026673001.1


In [7]:
%%bash
cd /hydra/neiro_bam
head -2 gffcmp.stringtie_merged.gtf.tmap

ref_gene_id	ref_id	class_code	qry_gene_id	qry_id	num_exons	FPKM	TPM		cov	len	major_iso_id	ref_match_len
SMESG000026673.1	SMEST026673001.1	=	MSTRG.1	SMEST026673001.1	2	0.000000	0.000000	0.000000	1072	SMEST026673001.1	553


In [9]:
%%bash
cd /hydra/neiro_bam
head -2 gffcmp.annotated.gtf

dd_Smes_g4_1	StringTie	transcript	20346	21503	.	+	.	transcript_id "SMEST026673001.1"; gene_id "MSTRG.1"; gene_name "SMESG000026673.1"; xloc "XLOC_000001"; ref_gene_id "SMESG000026673.1"; cmp_ref "SMEST026673001.1"; class_code "="; tss_id "TSS1";
dd_Smes_g4_1	StringTie	exon	20346	20485	.	+	.	transcript_id "SMEST026673001.1"; gene_id "MSTRG.1"; exon_number "1";


In [2]:
%%bash
cd /hydra/neiro_bam
grep -v -P '\t\.\t\.\t' stringtie_merged.gtf > stringtie_merged_R.gtf

In [1]:
28097-22129

5968

# 3. GO terms for Rink high-confidence transcripts

The GO terms for Rink's high-confidence transcripts were extracted with Interproscan. The GO terms of these transcripts were later compared to the new identified transcripts (1-3-Genome-annotation-analysis-R, 3. Gene ontology analysis of new coding transcripts).   

## 3.1 High-confidence transcripts from annotation file

The high-confidence transcripts were extracted from the annotation file.

In [ ]:
%%bash
cd /hydra/GO
gffread -x final_dd_Smed_g4_transcripts.fa -g ~/sexual_genome_annotation_files/sexual_genome_annotation/final_dd_Smed_g4.fa ~/sexual_genome_annotation_files/sexual_genome_annotation/smes_v2_hconf_SMESG.gff3

In [4]:
%%bash
cd /hydra/GO
gffread --version

0.12.1


In [1]:
%%bash
cd /hydra/GO
grep -c ">" final_dd_Smed_g4_transcripts.fa

31102


## 3.2 Transdecoder

The open reading frames were computed with Transdecoder.

In [ ]:
#%%bash
#cd /hydra/GO
#TransDecoder.LongOrfs -t final_dd_Smed_g4_transcripts.fa -o Transdecoder_final

In [5]:
%%bash
cd /hydra/GO
TransDecoder.LongOrfs --version

TransDecoder.LongOrfs 5.5.0


In [2]:
%%bash
cd /hydra/GO/Transdecoder_final
grep -c ">" longest_orfs.pep

36368


The last "\*" character of the peptide sequences was removed.

In [ ]:
%%bash
cd /hydra/GO/Transdecoder_final
cp longest_orfs.pep longest_orfs_ip.pep
sed -i 's_*__g' longest_orfs_ip.pep
head longest_orfs_ip.pep

## 3.3 Interproscan

The peptide file generated with Transdecoder was split up into parts for Interproscan:

In [ ]:
%%bash
cd /hydra/GO/Transdecoder_final
seqkit range -r 1:5000 longest_orfs_ip.pep > orfs_1_5k.fa.pep
seqkit range -r 5001:10000 longest_orfs_ip.pep > orfs_5k_10k.fa.pep
seqkit range -r 10001:15000 longest_orfs_ip.pep > orfs_10k_15k.fa.pep
seqkit range -r 15001:20000 longest_orfs_ip.pep > orfs_15k_20k.fa.pep
seqkit range -r 20001:25000 longest_orfs_ip.pep > orfs_20k_25k.fa.pep
seqkit range -r 25001:30000 longest_orfs_ip.pep > orfs_25k_30k.fa.pep
seqkit range -r 30001:36368 longest_orfs_ip.pep > orfs_30k_36k.fa.pep

The GO terms (and other peptide domain information) were obtained with Interproscan: 

In [ ]:
#%%bash
#cd /hydra/GO
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_1_5k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_5k_10k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_10k_15k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_15k_20k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_20k_25k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_25k_30k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &
#nohup ~/software/interproscan-5.39-77.0/interproscan.sh -i ./Transdecoder_final/orfs_30k_36k.fa.pep -d interpro -appl Pfam, SUPERFAMILY -goterms &

## 3.4 GO terms

The GO terms were extracted from the Interproscan output file and retabulated (interpro_rink.GO.tab). 

In [3]:
%%bash
cd /hydra/GO/interpro
ls *tsv

orfs_10k_15k.fa.pep.tsv
orfs_15k_20k.fa.pep.tsv
orfs_1_5k.fa.pep.tsv
orfs_20k_25k.fa.pep.tsv
orfs_25k_30k.fa.pep.tsv
orfs_30k_31k.fa.pep.tsv
orfs_5k_10k.fa.pep.tsv


In [ ]:
paths = ["../GO/interpro/orfs_1_5k.fa.pep.tsv", "../GO/interpro/orfs_5k_10k.fa.pep.tsv", "../GO/interpro/orfs_10k_15k.fa.pep.tsv", "../GO/interpro/orfs_15k_20k.fa.pep.tsv", "../GO/interpro/orfs_20k_25k.fa.pep.tsv", "../GO/interpro/orfs_25k_30k.fa.pep.tsv"]
transcripts = []
gos = []
for path in paths:
    f = open(path)
    content = f.readlines()
    print(len(content))
    for i in range(0, len(content)):
        if len(content[i].split("\t")) == 14 and content[i].split("\t")[13] != "\n":
            transcript = content[i].split("\t")[0].rpartition(".")[0]
            go = content[i].split("\t")[13].split("\n")[0].split("|")
            #transcripts were only included once
            if transcript not in transcripts:
                transcripts.append(transcript)
                gos.append(go)

In [ ]:
f = open("../GO/interpro/interpro_rink.GO.tab", "a+")
for i in range(0, len(transcripts)):
    f.write(str(transcripts[i])+"\t")
    for j in range(0, len(gos[i])):
        if j == (len(gos[i])-1):
            f.write(str(gos[i][j])+"\n")
        else:
            f.write(str(gos[i][j])+", ")

In [4]:
%%bash
cd /hydra/GO/interpro/
head -2 interpro_rink.GO.tab
wc -l interpro_rink.GO.tab

SMEST000228003.1	GO:0046983
SMEST026320001.1	GO:0003723
11194 interpro_rink.GO.tab


# 4. Coding and non-coding transcripts from new annotation

The transcripts from the new annotation were characterised.

## 4.1 Putative ncRNAs from annotation file

The gffcompare tmap file was imported:

In [1]:
import pandas as pd
gffcmp = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/gffcmp.stringtie_merged.gtf.tmap", sep="\t")

The total number of transcripts in the new annotation was calculated:

In [13]:
len(gffcmp)

91068

In [ ]:
The total number

In [10]:
len(gffcmp[gffcmp["class_code"] == "="])

28624

In [13]:
ncrna = gffcmp[gffcmp["class_code"].isin(['u', 'i', "x"])]

In [14]:
len(ncrna)

11255

In [52]:
len(gffcmp[gffcmp["class_code"].isin(['c', 'k', 'm', 'n', 'j', 'e', 'o'])])

50834

The IDs of the putative ncRNAs were saves as a list file:

In [57]:
#ncrna.iloc[:, 4].to_csv("../sexual_genome_annotation_files/ncrna_Neiro/ncrna_list.csv", sep ="\t", index=False)

The ncRNAs were extracted as a gtf file:

In [ ]:
#%%bash
#cd ../../../../../../data/ngs/smed/facs_rnaseq_datasets/sexual/neiro_bam
#grep -F -f ~/sexual_genome_annotation_files/ncrna_Neiro/ncrna_list.csv gffcmp.annotated.gtf > ~/sexual_genome_annotation_files/ncrna_Neiro/stringtie.annotated.ncrna.gtf

In [19]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro
head -2 stringtie.annotated.ncrna.gtf

dd_Smes_g4_1	StringTie	transcript	30235	31845	.	+	.	transcript_id "MSTRG.2.1"; gene_id "MSTRG.2"; xloc "XLOC_000002"; class_code "u"; tss_id "TSS4";
dd_Smes_g4_1	StringTie	exon	30235	30464	.	+	.	transcript_id "MSTRG.2.1"; gene_id "MSTRG.2"; exon_number "1";


The ncRNA sequences were extracted using the gtf file:

In [ ]:
#%%bash
#cd ~/sexual_genome_annotation_files/ncrna_Neiro
#gffread -w ncrna_transcripts.fa -g ../sexual_genome_annotation/final_dd_Smed_g4.fa stringtie.annotated.ncrna.gtf

In [21]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro
head -2 ncrna_transcripts.fa

>MSTRG.2.1
GTAAAGTTAACCAAAACTTGTATTTGTTTATAACAATTGTTTAAATAATGTCTACTTCATCAAAAATATT


In [22]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro
grep -c ">" ncrna_transcripts.fa

11361


Because of the grep function, some false postivies were included in the fasta file.

## 4.2 Transdecoder

### 4.2.1 Open reading frames

The open reading frames for new transcripts were computed with Transdecoder:

In [ ]:
#%%bash
#cd ~/sexual_genome_annotation_files/ncrna_Neiro
#TransDecoder.LongOrfs -t ncrna_transcripts.fa -m 10 -O Transdecoder_final

In [24]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
grep -c ">" longest_orfs.pep

144368


In [26]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
head -2 longest_orfs.gff3

MSTRG.2.1	transdecoder	gene	1	1569	.	-	.	ID=GENE.MSTRG.2.1~~MSTRG.2.1.p1;Name=ORF%20type%3Acomplete%20len%3A250%20%28-%29
MSTRG.2.1	transdecoder	mRNA	1	1569	.	-	.	ID=MSTRG.2.1.p1;Parent=GENE.MSTRG.2.1~~MSTRG.2.1.p1;Name=ORF%20type%3Acomplete%20len%3A250%20%28-%29


In [27]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
head -2 longest_orfs.pep

>MSTRG.2.1.p1 type:complete len:250 gc:universal MSTRG.2.1:1042-293(-)
MKGDDFNIITAPVPINRIWWYSLTNRQRIALVFYMSFYVAGTLTNTASMFIDKFYIYIMRLESLQMGSADPIDYKYLLEVQIVRGFWREDVHEVVDKVFRGKSIGYIKTNLMIPVEIWNNCQVRSFRGIPCHSVAIICLIFGMLILYYHCTTVALFRTFMILNANLAAILLFIAMSMEYSAAVEYDYCVNSVFMNRKTGGKAFVRGRYYNRTLEASGSTFKLMMVGDILFFCPMIGLGCYLLFCNRENL*


Transcripts without an open reading frame were extracted:

In [ ]:
import pandas as pd
orf10 = pd.read_csv("../sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/longest_orfs.gff3", sep="\t", header=None)

In [ ]:
non_orf = []
orfs = pd.Categorical(orf10.iloc[:,0]).categories
for i in range(len(ncrna_list)):
    if ncrna_list.iloc[i, 0] not in orfs:
          non_orf.append(ncrna_list.iloc[i, 0])

In [ ]:
pd.Series(non_orf).to_csv("../sexual_genome_annotation_files/ncrna_Neiro/noncoding10aa_list.csv"

In [28]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
wc -l noncoding10aa_list.csv

1 noncoding10aa_list.csv


### 4.2.2 Preparation for Interproscan

The peptide fasta file was split up into two parts: sequences longer than 100 amino acids and sequences shorter than 100 amino acids. 

In [ ]:
%%bash
cd ~/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
seqkit seq -m 100 longest_orfs.pep > aa100longer_merged.fa.pep

In [ ]:
%%bash
cd ~/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
seqkit seq -M 100 longest_orfs.pep > aa100shorter_merged.fa.pep

In [30]:
#%%bash
#cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
#head -2 aa100longer_merged.fa.pep

In [31]:
#%%bash
#cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
#head aa100shorter_merged.fa.pep

In [32]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
grep -c ">" aa100longer_merged.fa.pep

2771


In [33]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
grep -c ">" aa100shorter_merged.fa.pep

141661


In [36]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
#cp aa100longer_merged.fa.pep aa100longer_merged_ip.fa.pep
#sed -i 's_*__g' aa100longer_merged_ip.fa.pep
head -2 aa100longer_merged_ip.fa.pep

>MSTRG.2.1.p1 type:complete len:250 gc:universal MSTRG.2.1:1042-293(-)
MKGDDFNIITAPVPINRIWWYSLTNRQRIALVFYMSFYVAGTLTNTASMFIDKFYIYIMR


In [37]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
#cp aa100shorter_merged.fa.pep aa100shorter_merged_ip.fa.pep
#sed -i 's_*__g' aa100shorter_merged_ip.fa.pep
head -2 aa100shorter_merged_ip.fa.pep

>MSTRG.2.1.p2 type:complete len:47 gc:universal MSTRG.2.1:48-188(+)
MSTSSKIFLASSFGFCLGMFGFVHISQKLERDVDLFSLINIIFNLF


## 4.3 Interproscan

### 4.3.1 Sequences longer than 100 amino acids

Interproscan was run on sequences longer than 100 amino acids.

In [ ]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
echo "#!/bin/bash" > interproscan_more100aa_24082020.sh
echo "/hydra/software/interproscan-5.46-81.0/interproscan.sh -i /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/aa100longer_merged_ip.fa.pep -d /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa_24082020 -goterms " >> interproscan_more100aa_24082020.sh
chmod +x interproscan_more100aa_24082020.sh
less interproscan_more100aa_24082020.sh

In [ ]:
#%%bash
#cd ~/sexual_genome_annotation_files/ncrna_Neiro/
#sudo nohup ./interproscan_more100aa_24082020.sh &

### 4.3.2 Sequences shorter than 1000 amino acids

Interproscan was run on sequences shorter than 100 amino acids.

In [ ]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
grep -c ">" aa100shorter_merged_ip.fa.pep

In [ ]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
sudo nohup /hydra/software/interproscan-5.46-81.0/interproscan.sh -i /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/aa100shorter_merged_ip.fa.pep -d /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020 -goterms & 

# 4.4 Characterisation of transcripts

The new transcripts were characterised: 

### 4.4.1 Coding sequences longer than 100 amino acids

All new transcripts with an open reading frame longer than 100 amino acids were selected:

In [ ]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
ls

In [ ]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
grep ">" aa100longer_merged_ip.fa.pep > aa100longer_merged_ip.list.txt

In [5]:
import pandas as pd
aa100longer_orf_pd = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/aa100longer_merged_ip.list.txt", header=None)

In [6]:
aa100longer_orf = []
for i in range(0, len(aa100longer_orf_pd)):
    aa100longer_orf.append(aa100longer_orf_pd.iloc[i,0].split(" ")[0].split(">")[1].split(".p")[0])

In [7]:
aa100longer_orf = pd.Categorical(aa100longer_orf).categories

In [8]:
len(aa100longer_orf)

2077

In [9]:
import pandas as pd
ncrna_list = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/ncrna_list.csv", header=None)

In [10]:
aa100longer_final = ncrna_list[ncrna_list.iloc[:,0].isin(aa100longer_orf)]

In [11]:
len(aa100longer_final)

2002

In [ ]:
#import pandas as pd
#pd.Series(pd.Categorical(aa100longer_final.iloc[:,0]).categories).to_csv("../sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa/codingmore100aa.csv")

In [63]:
%%bash
wc -l /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa/codingmore100aa.csv

2002 /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa/codingmore100aa.csv


In total, there were 2002 new transcripts with an open reading frame longer than 100 amino acids.

In [12]:
import pandas as pd
interpro_more100aa_df = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa_24082020/aa100longer_merged_ip.fa.pep.tsv", sep="\t",  header=None)

In [13]:
interpro_more100aa_genes = []
for i in range(0, len(interpro_more100aa_df)):
    interpro_more100aa_genes.append(interpro_more100aa_df.iloc[i,0].rpartition(".")[0])
interpro_more100aa_df["gene"] = interpro_more100aa_genes
interpro_more100aa_genes = pd.Categorical(interpro_more100aa_genes).categories

In [14]:
len(interpro_more100aa_genes)

1320

In total, 1320 of the 2002 transcripts had an Interproscan hit and were thus classified as coding transcripts.

GO terms were assigned to the new coding sequences longer than 100 amino acids:

In [12]:
import pandas as pd
ncrna_list = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/ncrna_list.csv", header=None)

In [19]:
interpro_more100aa_GO_draft = interpro_more100aa_df[interpro_more100aa_df["gene"].isin(ncrna_list.iloc[:,0].to_list())]

In [29]:
interpro_more100aa_GO = interpro_more100aa_GO_draft[interpro_more100aa_GO_draft.iloc[:,13].notnull()]

In [31]:
interpro_more100aa_GO_genes = []
interpro_more100aa_GO_go = []
for i in range(0, len(interpro_more100aa_GO)):
    gene = interpro_more100aa_GO.iloc[i,0].rpartition(".")[0]
    gene2 = interpro_more100aa_GO.iloc[i,0]
    if gene not in interpro_more100aa_GO_genes:
        interpro_more100aa_GO_genes.append(gene)
        df = interpro_more100aa_GO[interpro_more100aa_GO.iloc[:,0] == gene2]
        gos_final = []
        for j in range(0, len(df)):
            gos = df.iloc[j, 13].split("|")
            for k in gos:
                if k not in gos_final:
                    gos_final.append(k)
        interpro_more100aa_GO_go.append(gos_final)

In [34]:
f = open("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa_24082020/interpro_more100aa.GO.May2021.tab", "a+")
for i in range(0, len(interpro_more100aa_GO_genes)):
    f.write(str(interpro_more100aa_GO_genes[i])+"\t")
    for j in range(0, len(interpro_more100aa_GO_go[i])):
        f.write(str(interpro_more100aa_GO_go[i][j])+", ")
        if j == (len(interpro_more100aa_GO_go[i])-1):
            f.write(str(interpro_more100aa_GO_go[i][j])+"\n")

In [11]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa_24082020/
wc -l interpro_more100aa.GO.May2021.tab

552 interpro_more100aa.GO.May2021.tab


### 4.4.2 Coding sequences shorter than 100 amino acids

All new transcripts with an open reading frame shorter than 100 amino acids were selected:

In [39]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/
grep ">" aa100shorter_merged_ip.fa.pep > aa100shorter_merged_ip.list.txt

In [15]:
import pandas as pd
aa100shorter_orf_pd = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/Transdecoder_final/aa100shorter_merged_ip.list.txt", header=None)

In [16]:
aa100shorter_orf = []
for i in range(0, len(aa100shorter_orf_pd)):
    aa100shorter_orf.append(aa100shorter_orf_pd.iloc[i,0].split(" ")[0].split(">")[1].split(".p")[0])

In [17]:
aa100shorter_orf = pd.Categorical(aa100shorter_orf).categories

In [18]:
len(aa100shorter_orf)

11357

In [20]:
aa100shorter_final = ncrna_list[ncrna_list.iloc[:,0].isin(aa100shorter_orf)]

In [49]:
import pandas as pd
pd.Series(pd.Categorical(aa100shorter_final.iloc[:,0]).categories).to_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/codingless100aa.csv")

In [50]:
%%bash
wc -l /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/codingless100aa.csv

11252 /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/codingless100aa.csv


In [22]:
import pandas as pd
interpro_less100aa_df = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/aa100shorter_merged_ip.fa.pep.tsv", sep="\t",  header=None)

In [23]:
interpro_less100aa_genes = []
for i in range(0, len(interpro_less100aa_df)):
    interpro_less100aa_genes.append(interpro_less100aa_df.iloc[i,0].rpartition(".")[0])
interpro_less100aa_df["gene"] = interpro_less100aa_genes
interpro_less100aa_genes = pd.Categorical(interpro_less100aa_genes).categories

In [24]:
len(interpro_less100aa_genes)

4408

In [25]:
interpro_less100aa_genes = interpro_less100aa_genes[~interpro_less100aa_genes.isin(interpro_more100aa_genes)]

In [26]:
len(interpro_less100aa_genes)

3841

In total, 3841 transcripts of the 11252 had an interpro hit and were classified as coding, and were not present in the longer sequences already.

GO terms were assigned to the new coding sequences shorter than 100 amino acids:

In [53]:
interpro_less100aa_GO_draft = interpro_less100aa_df[interpro_less100aa_df["gene"].isin(ncrna_list.iloc[:,0].to_list())]

In [58]:
interpro_less100aa_GO = interpro_less100aa_GO_draft[interpro_less100aa_GO_draft.iloc[:,13].notnull()]

In [62]:
interpro_less100aa_GO = interpro_less100aa_GO[interpro_less100aa_GO["gene"].isin(interpro_less100aa_genes)]

In [63]:
interpro_less100aa_GO_genes = []
interpro_less100aa_GO_go = []
for i in range(0, len(interpro_less100aa_GO)):
    gene = interpro_less100aa_GO.iloc[i,0].rpartition(".")[0]
    gene2 = interpro_less100aa_GO.iloc[i,0]
    if gene not in interpro_less100aa_GO_genes:
        interpro_less100aa_GO_genes.append(gene)
        df = interpro_less100aa_GO[interpro_less100aa_GO.iloc[:,0] == gene2]
        gos_final = []
        for j in range(0, len(df)):
            gos = df.iloc[j, 13].split("|")
            for k in gos:
                if k not in gos_final:
                    gos_final.append(k)
        interpro_less100aa_GO_go.append(gos_final)

In [64]:
f = open("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/interpro_less100aa.GO.May2021.tab", "a+")
for i in range(0, len(interpro_less100aa_GO_genes)):
    f.write(str(interpro_less100aa_GO_genes[i])+"\t")
    for j in range(0, len(interpro_less100aa_GO_go[i])):
        f.write(str(interpro_less100aa_GO_go[i][j])+", ")
        if j == (len(interpro_less100aa_GO_go[i])-1):
            f.write(str(interpro_less100aa_GO_go[i][j])+"\n")

In [66]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/
wc -l interpro_less100aa.GO.May2021.tab

179 interpro_less100aa.GO.May2021.tab


Merging the GO table for all new coding transcripts:

In [1]:
%%bash
cd /hydra/GO/interpro/
cat /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa_24082020/interpro_more100aa.GO.May2021.tab /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/interpro_less100aa.GO.May2021.tab > ncrna_new_coding_May2021.GO.tab

Merging the GO table for all coding transcripts:

In [69]:
%%bash
cd /hydra/GO/interpro/
cat /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_more100aa_24082020/interpro_more100aa.GO.May2021.tab /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/interpro_less100aa.GO.May2021.tab interpro_rink.GO.tab > ncrna_coding_May2021.GO.tab

In [2]:
%%bash
cd /hydra/GO/interpro/
wc -l ncrna_new_coding_May2021.GO.tab

731 ncrna_new_coding_May2021.GO.tab


### 4.4.3 Non-coding transcripts

The list of non-coding transcripts was defined as the complement of the new coding transcripts from the previous section. 

In [116]:
import pandas as pd
ncrna_list = pd.read_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/ncrna_list.csv", header=None)

In [117]:
noncoding = ncrna_list[~ncrna_list.iloc[:,0].isin(interpro_more100aa_genes)]
noncoding = noncoding[~noncoding.iloc[:,0].isin(interpro_less100aa_genes)]

In [27]:
coding = ncrna_list[ncrna_list.iloc[:,0].isin(interpro_more100aa_genes)|ncrna_list.iloc[:,0].isin(interpro_less100aa_genes)]

The non-coding transcripts were saved as a csv file.

In [118]:
noncoding.to_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/noncoding.csv", index=False, header=None)

In [119]:
%%bash
wc -l /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/noncoding.csv

6190 /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/noncoding.csv


In [28]:
coding.to_csv("/hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/coding_new.csv", index=False, header=None)

In [29]:
%%bash
wc -l /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/coding_new.csv

5065 /hydra/sexual_genome_annotation_files/ncrna_Neiro/interproscan_less100aa_24082020/coding_new.csv


The non-coding transcripts with the non-codiing transcripts without ORFs:

In [120]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
cat noncoding10aa_list.csv interproscan_less100aa_24082020/noncoding.csv > noncoding_final_May2021.csv

In [5]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
grep -f noncoding_final_May2021.csv stringtie_merged_exp.gtf > noncoding_final_May2021.gtf

In [1]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
grep -f interproscan_less100aa_24082020/coding_new.csv stringtie_merged_exp.gtf > coding_final_May2021.gtf

In [6]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
wc -l noncoding_final_May2021.gtf

65237 noncoding_final_May2021.gtf


In [2]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
wc -l coding_final_May2021.gtf

65182 coding_final_May2021.gtf


In [8]:
%%bash
cd /hydra/sexual_genome_annotation_files/ncrna_Neiro/
wc -l noncoding_final_May2021.csv

6191 noncoding_final_May2021.csv


The final frequencies are:
* All transcripts: 91068
* Identical to Rink: 28624
* Novel coding: 5161
* Novel non-coding: 6191

# FINNISHED